src/circuit/gadget/poseidon/pow5.rs

```diff
@@ -438,83 +427,73 @@ impl<F: FieldExt> From<CellValue<F>> for StateWord<F> {
         self,
         region: &mut Region<F>,
-        config: &Pow5T3Config<F>,
+        config: &Pow5Config<F, WIDTH, RATE>,
         round: usize,
         offset: usize,
     ) -> Result<Self, Error> {
         Self::round(region, config, round, offset, config.s_full, |_| {
-            let q_0 = self.0[0]
-                .value
-                .map(|v| v + config.round_constants[round][0]);
-            let q_1 = self.0[1]
-                .value
-                .map(|v| v + config.round_constants[round][1]);
-            let q_2 = self.0[2]
-                .value
-                .map(|v| v + config.round_constants[round][2]);
-
-            let r_0 = q_0.map(|v| v.pow(&config.alpha));
-            let r_1 = q_1.map(|v| v.pow(&config.alpha));
-            let r_2 = q_2.map(|v| v.pow(&config.alpha));
-
+            let q = self
+                .0
+                .iter()
+                .enumerate()
+                .map(|(idx, word)| word.value.map(|v| v + config.round_constants[round][idx]));
+            let r: Option<Vec<F>> = q.map(|q| q.map(|q| q.pow(&config.alpha))).collect();
             let m = &config.m_reg;
-            let r = r_0.and_then(|r_0| r_1.and_then(|r_1| r_2.map(|r_2| [r_0, r_1, r_2])));
-
-            Ok((
-                round + 1,
-                [
-                    r.map(|r| m[0][0] * r[0] + m[0][1] * r[1] + m[0][2] * r[2]),
-                    r.map(|r| m[1][0] * r[0] + m[1][1] * r[1] + m[1][2] * r[2]),
-                    r.map(|r| m[2][0] * r[0] + m[2][1] * r[1] + m[2][2] * r[2]),
-                ],
-            ))
+            let state = m.iter().map(|m_i| {
+                r.as_ref().map(|r| {
+                    r.iter()
+                        .enumerate()
+                        .fold(F::zero(), |acc, (j, r_j)| acc + m_i[j] * r_j)
+                })
+            });
+
+            Ok((round + 1, state.collect::<Vec<_>>().try_into().unwrap()))
         })
     }
 
-    fn partial_round(
+    fn partial_round<const RATE: usize>(
         self,
         region: &mut Region<F>,
-        config: &Pow5T3Config<F>,
+        config: &Pow5Config<F, WIDTH, RATE>,
         round: usize,
         offset: usize,
     ) -> Result<Self, Error> {
         Self::round(region, config, round, offset, config.s_partial, |region| {
             let m = &config.m_reg;
-
-            let p = self.0[0].value.and_then(|p_0| {
-                self.0[1]
-                    .value
-                    .and_then(|p_1| self.0[2].value.map(|p_2| [p_0, p_1, p_2]))
-            });
-
-            let r = p.map(|p| {
-                [
-                    (p[0] + config.round_constants[round][0]).pow(&config.alpha),
-                    p[1] + config.round_constants[round][1],
-                    p[2] + config.round_constants[round][2],
-                ]
+            let p: Option<Vec<_>> = self.0.iter().map(|word| word.value).collect();
+
+            let r: Option<Vec<_>> = p.map(|p| {
```

Comment:
Again I don't particularly like the shadowing in `p.map(|p| ...)`.

Reply:
I can try and address this in the rewrite.